File: leetcode/cpp/1593.split-a-string-into-the-max-number-of-unique-substrings.cpp
class Solution {
public:
    int maxUniqueSplit(string s) {
        size_t ans = 0;
        unordered_set<string> set;
        backtrack(0, std::move(set), s, ans);
        return ans;
    }
private:
    void backtrack(int i, unordered_set<string>&& set, const string& s, size_t& ans) {
        if (i == s.length()) {
            ans = max(ans, set.size());
            return;
        }
        for (int j = 1; j <= s.length()-i; j++) {
            const string substr = s.substr(i, j);
            if (set.contains(substr)) {
                continue;
            }
            set.insert(substr);
            backtrack(i + j, std::move(set), s, ans);
            set.erase(substr);
        }
    }
};

File: leetcode/python/1593.split-a-string-into-the-max-number-of-unique-substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        ans = 0
        def backtrack(i: int, uniq) -> int:
            if i == len(s):
                nonlocal ans
                ans = max(ans, len(uniq))
                return
            for j in range(i+1, len(s)+1):
                if s[i:j] in uniq:
                    continue
                uniq.add(s[i:j])
                backtrack(j, uniq)
                uniq.remove(s[i:j])
        uniq = set()
        backtrack(0, uniq)
        return ans

